sendrecv/gst-rust/src/main.rs

```diff
@@ -0,0 +1,480 @@
+        })
+        .unwrap();
+    pipe.clone()
+        .dynamic_cast::<gst::Bin>()
+        .unwrap()
+        .add(&decodebin)
+        .unwrap();
+    decodebin.sync_state_with_parent().unwrap();
+    webrtc.link(&decodebin).unwrap();
+    None
+}
+
+fn send_ice_candidate_message(
+    app_control: &Arc<Mutex<AppControl>>,
+    values: &[glib::Value],
+) -> Option<glib::Value> {
+    let app_control = app_control.lock().unwrap();
+    if app_control.app_state < AppState::PeerCallNegotiating {
+        panic!("Can't send ICE, not in call");
+    }
+
+    let _webrtc = values[0].get::<gst::Element>().expect("Invalid argument");
+    let mlineindex = values[1].get::<u32>().expect("Invalid argument");
+    let candidate = values[2].get::<String>().expect("Invalid argument");
+    let message = json!({
+          "ice": {
+            "candidate": candidate,
+            "sdpMLineIndex": mlineindex,
+          }
+        });
+    app_control.ws_sender.send(message.to_string()).unwrap();
+    None
+}
+
+fn start_pipeline(app_control: &Arc<Mutex<AppControl>>) -> Result<gst::Element, Error> {
+    let pipe = gst::parse_launch(&format!(
+        "webrtcbin name=sendrecv {}
+        videotestsrc pattern=ball ! videoconvert ! queue ! vp8enc deadline=1 ! rtpvp8pay !
+        queue ! {}96 ! sendrecv.
+        audiotestsrc wave=red-noise ! audioconvert ! audioresample ! queue ! opusenc ! rtpopuspay !
+        queue ! {}97 ! sendrecv.
+        ",
+        STUN_SERVER, RTP_CAPS_VP8, RTP_CAPS_OPUS
+    ))?;
+
+    let webrtc = pipe.clone()
+        .dynamic_cast::<gst::Bin>()
+        .unwrap()
+        .get_by_name("sendrecv")
+        .unwrap();
+    let app_control_clone = app_control.clone();
+    webrtc.connect("on-negotiation-needed", false, move |values| {
+        on_negotiation_needed(&app_control_clone, values)
+    })?;
+
+    let app_control_clone = app_control.clone();
+    webrtc.connect("on-ice-candidate", false, move |values| {
+        send_ice_candidate_message(&app_control_clone, values)
+    })?;
+
+    let pipe_clone = pipe.clone();
+    // TODO: replace with webrtc.connect_pad_added
+    webrtc.connect("pad-added", false, move |values| {
+        on_incoming_stream(values, &pipe_clone)
+    })?;
+
+    pipe.set_state(gst::State::Playing).into_result()?;
+
+    return Ok(webrtc);
+}
+
+struct WsClient {
+    webrtc: Option<gst::Element>,
+    app_control: Arc<Mutex<AppControl>>,
+}
+
+struct AppControl {
+    app_state: AppState,
+    ws_sender: ws::Sender,
+    peer_id: String,
+}
+
+impl WsClient {
+    fn update_state(&self, state: AppState) {
+        self.app_control.lock().unwrap().app_state = state
+    }
+}
+
+impl ws::Handler for WsClient {
+    fn on_open(&mut self, _: ws::Handshake) -> ws::Result<()> {
+        self.update_state(AppState::ServerConnected);
+        self.update_state(register_with_server(&self.app_control.clone()));
+        Ok(())
+    }
+
+    fn on_message(&mut self, msg: ws::Message) -> ws::Result<()> {
```

Comment:
Is there also some kind of `on_error` or so that would be called if the connection attempt fails or anything else goes wrong?